### ISSUE TITLE
Reducing bamsurgeon runtime.

### ISSUE THREAD

**Issue by BrianOSullivanGit** [adamewing/bamsurgeon#170]
Hello,

Firstly, thanks for a very useful tool for somatic variant simulation.

As part of our simulation work we are using bamsurgeon (version 1.2) to try to create a virtual tumour BAM containing a large number (>20K) of low frequency somatic SNVs. The bamsurgeon run times required given our current compute resources are proving prohibitive.

In an effort to improve runtime I looked into the '--ignorepileup' option.
When set, if I understand correctly, this option restricts the pileup check to the somatic SNV target locus, as opposed to across the entire pileup region. Setting this option significantly improved runtime.

However, running with '--ignorepileup' (please let me know if my understanding is incorrect) risks inserting the same somatic SNV in both haplotypes. In order to prevent this we are considering using phased bam files as input. Could you confirm if my understanding of the consequences of running with the '--ignorepileup' are correct?

Finally, during the run, bamsurgeon creates individual BAM files containing the pileup (including reads that have been  mutated) for each target somatic SNV locus. These are in turn fed to bwa aln to create sai index files for both reads and mates before again calling bwa again to generate a re-mapped SAM of the pileup.

Is it possible to setup bamsurgeon to aggregate these pileup BAMs together before calling bwa?
So, rather than individual calls to bwa for each variant to be inserted, the call can be made just once at the end, thereby reducing runtime?

Thanks and regards, Brian

**Comment by adamewing** [comment]
Hi Brian, Glad it's useful for you - you're right that bamsurgeon isn't particularly efficient. You're correct that `--ignorepileup` will cause interference with haplotypes. In general though, bamsurgeon is not haplotype aware; it avoids affecting haplotypes by avoiding making mutations that share reads with SNPs, depending on the parameters used. So if haplotypes are important in your use case (they should be in pretty much every case) you should definintely use single-phase .bams and merge them after mutation.

With the way bamsurgeon/addsnv is structured, it isn't currently possible to aggregate BAMs before realigning - unfortunately that would require some non-trivial refactoring. That said, a re-write of all of this may be in the cards for the future. I imagine you're taking advantage of the `-p/--procs` option to spread the realignment jobs over multiple processes. If you're using bwa you could try using `bwa shm` to load the index into shared memory before running addsnv. The other thing I would mention is, because you mentioned `bwa aln`, make sure `bwa aln` is really what you want to be using. Most folks are using bwa mem, minimap2, etc and merging realigned reads into a .bam initially generated with a different aligner will introduce artifacts.

**Comment by BrianOSullivanGit** [comment]
Hi Adam,

Sorry for my delay in responding. Thanks a lot, your reply is very helpful.
I wasn't aware of the shared mem option with bwa. I hope taking out those disc access should speed things up all right.
Our cluster is max'd out at the moment. As soon as I get back on there I will update this note with how much it speeds it up.

Thanks for pointing out the need to use the same aligner with bamsurgeon as was used on the original bam. That makes complete sense, however, I had overlooked it. From looking at the sam headers they were indeed put together with bwa mem so I will go that route.

Yes, I use the -procs option which does improve runtime. Am not sure why but I noticed that different -proc settings can also produce slight differences in allele frequencies as rounding comes into play. For example, with "len(readlist): 42" and selected "VAF: 0.045455", setting "-procs 8" for me will round up to "mutated: 2", whereas "-procs 8"  will result in "mutated: 1". It doesn't really cause a problem in our case but it is something to be aware of (apologies if it has been noted somewhere previously).

Thanks again for your help,

Regards, Brian

**Comment by adamewing** [comment]
OK, great. Are you using the `--seed` option to set a RNG seed? If not that's probably where the stochastic read counts are coming from. --Adam

**Comment by BrianOSullivanGit** [comment]
Hi Adam,

Thanks, seed did the trick there, my mistake.

On a separate note, when I run addsnv.py with '--aligner mem' and '--procs 16' output flags a htslib error (E::idx_find_and_load). Do you know is this error significant (/can I ignore it)?
(it nearly looks like one process is trying to access the index before another has been created)


INFO 2021-01-11 16:12:24,198 haplo_1_10132238_10132238 len(readlist): 62
INFO 2021-01-11 16:12:24,198 haplo_1_10132238_10132238 selected VAF: 0.444444
INFO 2021-01-11 16:12:24,199 haplo_1_10132238_10132238 picked: 28
INFO 2021-01-11 16:12:24,200 haplo_1_10132238_10132238 wrote: 62, mutated: 28
[E::idx_find_and_load] Could not retrieve index file for 'addsnv.tmp/haplo_1_10132238_10132238.tmpbam.de3ba933-022a-4ca8-8565-78552d63ca9a.bam'
[E::idx_find_and_load] Could not retrieve index file for 'addsnv.tmp/haplo_1_10132238_10132238.tmpbam.de3ba933-022a-4ca8-8565-78552d63ca9a.bam'
:
:
INFO 2021-01-11 16:12:25,832 haplo_1_10132238_10132238 indexing: samtools index addsnv.tmp/haplo_1_10132238_10132238.tmpbam.de3ba933-022a-4ca8-8565-78552d63ca9a.bam
INFO 2021-01-11 16:12:25,855 haplo_1_10132238_10132238 removing addsnv.tmp/haplo_1_10132238_10132238.tmpbam.de3ba933-022a-4ca8-8565-78552d63ca9a.fastq

I am running with the following version of pysam,

[bosullivan@compute01 TEST_CASE]$ python
Python 3.7.0 (default, Jun 28 2018, 13:15:42) 
[GCC 7.2.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pysam
>>> pysam.__version__
'0.16.0.1'
>>> 

Would this error notification be an underlying pysam issue?
ie.,
https://github.com/pysam-developers/pysam/issues/939

(should I upgrade/downgrade my pysam version)

Thanks again for your help!

**Comment by BrianOSullivanGit** [comment]
Just to add for ref., my command line is,

[bosullivan@lugh TEST_CASE]$ head  bamsurgeon-seed1234-force-procs16-mem-picardjar_bwa_shm_B.log
INFO 2021-01-11 18:34:02,455 starting /home/bosullivan/anaconda3/bin/addsnv.py called with args: /home/bosullivan/anaconda3/bin/addsnv.py --seed 1234 --force --procs 16 --aligner mem --picardjar /home/bosullivan/bin/picard-tools-1.131/picard.jar --minmutreads 0 -v test.cfg -f /data/bosullivan/TMB/TCGA-ACC/TCGA-P6-A5OH-11A-01D-A30A-10_Illumina.bam -r /data/bosullivan/TMB/REF/NEW_REF/GRCh37-lite.fa -o test_tumour.bam

And to confirm, this error is not flag'd with 'bwa aln' etc..

**Comment by BrianOSullivanGit** [comment]
last quick update....
I built the output bam and had a look at the pileup at the site in question.
The 28 alt reads are there, so hopefully can disregard the error (?). Thanks.

[bosullivan@compute01 TEST_CASE]$ samtools mpileup --min-BQ 0 -r 1:10132238-10132238 -f /data/bosullivan/TMB/REF/NEW_REF/GRCh37-lite.fa test_tumour.bam 
[mpileup] 1 samples in 1 input files
1       10132238        C       68      A,,a.A.AA.,Aa..A,a....aaAA..,a,.aa.,aaa..A,aaa,.a,,.,a,..a,A,,a,,,,,      e@GFln?qjg>pGfIp>HgolF!G?l@DGEHGCGEGG!!Ffo!F!!HE!!!E#!!EE!#D#!9#!!!!

**Comment by adamewing** [comment]
Yes you can ignore that - it's an HTSlib issue with an overzealous warning. The issue you linked to has details and a workaround via pysam.set_verbosity() (I've done this in other projects but haven't updated bamsurgeon with the workaround yet).

**Comment by BrianOSullivanGit** [comment]
Hi,

Sorry for the delay in updating this. I wanted to post some results showing the time improvements made as a result of the optimisations advised above but I do not currently have cluster access. In the last small scale test I did, creation times for BAMs for each spiked in SNVs went from about 4-5 second/BAM to about 1 second.  Thanks a lot for these significant improvements. Closing case.

Best regards, Brian
